Allow disable automated cleaning

fmuyassarov opened the pull request:
Disabling [disk cleaning](https://github.com/metal3-io/metal3-docs/blob/master/design/cluster-api-provider-metal3/allow_disabling_node_disk_cleaning.md) implementation.
Fixes: https://github.com/metal3-io/baremetal-operator/issues/626
Related changes in CAPM3: https://github.com/metal3-io/cluster-api-provider-metal3/pull/156

fmuyassarov commented:
/cc @hardys 
/cc @dhellmann 
/cc @maelk

fmuyassarov commented:
/cc @dtantsur

hardys commented on config/crd/bases/metal3.io_baremetalhosts.yaml:
```diff
@@ -121,6 +121,10 @@ spec:
               description:
                 description: Description is a human-entered text used to help identify the host
                 type: string
+              disableAutomatedClean:
```
I know the design for this landed already, but are we OK with using an annotation for disabling inspection ref metal3-io/metal3-docs#155 and #787, but disabling cleaning via the BMH spec?

hardys replied:
@dhellmann any thoughts on that?  I'm not blocking on it, just wondering if we have any guidelines on what should be in the spec vs annotations?

In practice it seems like annotations are used to modify the baremetalhost_controller behavior, but the spec is used to influence the provisioner configuration, but that seems like an implementation detail rather than an API design pattern :)

dhellmann replied:
I'm not sure why we used an annotation for disabling inspection. Maybe because it wasn't really a thing we thought most users would (or should) do? Or because the semantics of changing an API field to enable inspection after a host was provisioned were confusing?

An API field to turn off cleaning seems reasonable, and the semantics of changing it later are clearer (at least to me). I could be convinced that the inspection and cleaning features should be treated the same way, as API fields.

I should have provided this feedback on the design review, but I think we don't want to use a boolean field. It's harder to manage changes in behavior. For example, today cleaning is all or none. In the future, it may be possible to clean some disks and leave others alone. We may want a 3rd mode to cover that case. Creating this field as an enum now would let us make that change relatively easily, without an API version change and upgrade process. So, I suggest a name like `automatedCleaningMode` with values `Full` and `None` instead of a bool here.

hardys replied:
> I'm not sure why we used an annotation for disabling inspection.

I did it that way to align with the re-introspection proposal that was already accepted : https://github.com/metal3-io/metal3-docs/blob/master/design/baremetal-operator/re-introspection-interface.md

> Maybe because it wasn't really a thing we thought most users would (or should) do? 

This is probably also true, the main use-case we have for it atm is when booting a live-iso, which likely isn't the most common path, and we might encourage folks to use a visible API in the spec for the wrong reasons (e.g faster deployments, but the same could be said for this API I guess)

dhellmann replied:
> > I'm not sure why we used an annotation for disabling inspection.
> 
> I did it that way to align with the re-introspection proposal that was already accepted : https://github.com/metal3-io/metal3-docs/blob/master/design/baremetal-operator/re-introspection-interface.md

Ah, good point. In that case the annotation is removed after the re-inspection is done, right? I suppose we could have had the operator toggle a flag that's part of the schema instead. But we probably based the re-inspection thing on the reboot API, and that was also based on an annotation.

> > Maybe because it wasn't really a thing we thought most users would (or should) do?
> 
> This is probably also true, the main use-case we have for it atm is when booting a live-iso, which likely isn't the most common path, and we might encourage folks to use a visible API in the spec for the wrong reasons (e.g faster deployments, but the same could be said for this API I guess)

Yeah, true. I have seen recommendations for using annotations for interactions between components (such as an annotation on a host that is consumed by something other than the BMO) or as early implementations of APIs that aren't ready to be locked into a CRD. That last use feels vague and fuzzy, but I think that's how I justified the reboot, re-inspection, and skip-inspection changes in my mind. Eventually, those *should* become part of the spec fields, when we figure out how to do that. This change seems to fit easily enough into the API fields that we can move right to that implementation. But I'll go with the group's consensus, so if others would be more comfortable using an annotation, that's OK with me.

fmuyassarov replied:
> > > I'm not sure why we used an annotation for disabling inspection.
> > 
> > 
> > I did it that way to align with the re-introspection proposal that was already accepted : https://github.com/metal3-io/metal3-docs/blob/master/design/baremetal-operator/re-introspection-interface.md
> 
> Ah, good point. In that case the annotation is removed after the re-inspection is done, right? 

That's right. The implementation PR is not merged yet though https://github.com/metal3-io/baremetal-operator/pull/607

>I suppose we could have had the operator toggle a flag that's part of the schema instead. But we probably based the re-inspection thing on the reboot API, and that was also based on an annotation.
> 
That's true. Re-inspection was designed to be similar interface as reboot API.

fmuyassarov replied:
> I should have provided this feedback on the design review, but I think we don't want to use a boolean field. It's harder to manage changes in behavior. For example, today cleaning is all or none. In the future, it may be possible to clean some disks and leave others alone. We may want a 3rd mode to cover that case. Creating this field as an enum now would let us make that change relatively easily, without an API version change and upgrade process. So, I suggest a name like `automatedCleaningMode` with values `Full` and `None` instead of a bool here. We can update the design doc after the code is approved.

@dhellmann Thanks, I agree with your point here and can change the both code + proposal. But are we okay to have enum here considering Ironic `Node.AutomatedClean` is bool. I know `automatedCleaningMode` doesn't have to be the same type as  `Node.AutomatedClean`, but just asking if there could be any issue with that.

hardys replied:
> But I'll go with the group's consensus, so if others would be more comfortable using an annotation, that's OK with me.

+1 I'm fine with this going into the spec, I just wanted to ensure we're all OK with that approach :)

And I agree that in the case of the inspection API it's fine to use an annotation for consistency (it'd be confusing to have both annotations and spec fields modifying that behavior IMO), and that's definitely an API which might see further changes in future, so it makes sense to avoid committing to a spec API now.

dhellmann replied:
> @dhellmann Thanks, I agree with your point here and can change the both code + proposal. But are we okay to have enum here considering Ironic `Node.AutomatedClean` is bool. I know `automatedCleaningMode` doesn't have to be the same type as `Node.AutomatedClean`, but just asking if there could be any issue with that.

I think it's OK for the APIs to be different. We've never made the assumption that metal3 users are Ironic users, and we try to design our APIs following kubernetes standards and conventions.

hardys commented on pkg/provisioner/ironic/ironic.go:
```diff
@@ -1212,6 +1212,37 @@ func (p *ironicProvisioner) Provision(hostConf provisioner.HostConfigData) (resu
 		return transientError(provisioner.NeedsRegistration)
 	}
 
+	p.log.Info("WARNING: automated cleaning is disabled for host",
```
Is this actually a warning?  Seems like it's just Info logging, since the behavior is what was requested?

In #787 I did also generate an event though, so it's visible via the BMH `describe` output, we may want the same in this case?

kashifest replied:
Shouldn't this logging be inside some check?

fmuyassarov replied:
@hardys , @kashifest agree with both points. Updating. Thanks.

hardys commented on pkg/provisioner/ironic/ironic.go:
```diff
@@ -1377,6 +1408,38 @@ func (p *ironicProvisioner) Deprovision(force bool) (result provisioner.Result,
 		return transientError(provisioner.NeedsRegistration)
 	}
 
+	p.log.Info("WARNING: automated cleaning is disabled for host",
```
Seems like we should refactor this into a function, given that the Provision/Deprovision code is the same?

kashifest replied:
Same here . The log is coming before the check

furkatgofurov7 replied:
+1 for refactoring it into a function.

kashifest reviewed (changes requested):
Looks Good. Minor comments in line. +1 for the idea of moving the code peace into a function

fmuyassarov commented:
Hi @dhellmann @hardys @kashifest 
Thanks for the reviews once more. I have updated the code as requested and opened a PR [#166](https://github.com/metal3-io/metal3-docs/pull/166) to update proposal in metal3-docs. PTAL.

fmuyassarov commented:
@dhellmann Please have look, updated as per the latest changes in the proposal.

dhellmann commented on apis/metal3.io/v1alpha1/baremetalhost_types.go:
```diff
@@ -278,6 +278,14 @@ type BareMetalHostSpec struct {
+	// +kubebuilder:default:=Full
+	// +kubebuilder:validation:Enum:=Full;Disabled
+	// +kubebuilder:validation:Optional
+	AutomatedCleaningMode string `json:"automatedCleaningMode,omitempty"`
```
For `BootMode` we define a type and set the kubebuilder options on that type. Can we do the same for this enum? That would let us have `const` values for comparisons, too.

fmuyassarov replied:
Yes of course. Fixed.

dhellmann commented on pkg/provisioner/ironic/ironic.go:
```diff
@@ -353,6 +353,44 @@ func (p *ironicProvisioner) findExistingHost() (ironicNode *nodes.Node, err erro
+func (p *ironicProvisioner) updateAutomatedClean(ironicNode *nodes.Node) (err error) {
+	var updates nodes.UpdateOpts
+
+	if p.host.Spec.AutomatedCleaningMode == "Disabled" {
```
Maybe this logic can be collapsed so that we're only creating the `updates` instance in one place and the conditional only sets the `Value` argument?

fmuyassarov replied:
That's a good point. Fixed it.

maelk commented on apis/metal3.io/v1alpha1/baremetalhost_types.go:
```diff
@@ -278,8 +278,25 @@ type BareMetalHostSpec struct {
+	// +optional
+	// +kubebuilder:default:=Full
+	// +kubebuilder:validation:Optional
+	AutomatedCleaningMode AutomatedCleaningMode `json:"automatedCleaningMode,omitempty"`
```
you could make this a pointer to simplify figuring out if it is omitted

fmuyassarov replied:
Added a check for nil pointer rather as you suggested. Thanks.

maelk commented on pkg/provisioner/ironic/ironic.go:
```diff
@@ -1212,6 +1245,21 @@ func (p *ironicProvisioner) Provision(hostConf provisioner.HostConfigData) (resu
 		return transientError(provisioner.NeedsRegistration)
 	}
 
+	if p.host.Spec.AutomatedCleaningMode == "None" {
```
What is this "None" ? You should also consider the empty string (or nil pointer if you make it a pointer)

fmuyassarov replied:
Oh sorry, this should be `Disabled`. Forgot to replace it.

fmuyassarov replied:
Fixed it. Thanks.

maelk commented on pkg/provisioner/ironic/ironic.go:
```diff
@@ -1377,6 +1425,21 @@ func (p *ironicProvisioner) Deprovision(force bool) (result provisioner.Result,
 		return transientError(provisioner.NeedsRegistration)
 	}
 
+	if p.host.Spec.AutomatedCleaningMode == "None" {
```
same here

fmuyassarov commented:
@dhellmann @maelk I think I addressed all the comments now. Please take a look once more. Thank you.

dhellmann commented on apis/metal3.io/v1alpha1/baremetalhost_types.go:
```diff
@@ -278,8 +278,25 @@ type BareMetalHostSpec struct {
+
+// Allowed automated cleaning modes
+const (
+	Disabled AutomatedCleaningMode = "Disabled"
```
It would be good to namespace these constants a bit like we do with the state machine states. I know we didn't do that with the checksum types just below, but we probably should have.

```suggestion
	CleaningModeDisabled AutomatedCleaningMode = "Disabled"
```

fmuyassarov replied:
Good point, fixed it.

dhellmann commented on apis/metal3.io/v1alpha1/baremetalhost_types.go:
```diff
@@ -278,8 +278,25 @@ type BareMetalHostSpec struct {
+// Allowed automated cleaning modes
+const (
+	Disabled AutomatedCleaningMode = "Disabled"
+	Full     AutomatedCleaningMode = "Full"
```
```suggestion
	CleaningModeFull     AutomatedCleaningMode = "Full"
```

dhellmann reviewed (commented):
One nit about names for the constants, since they're public. Otherwise this looks good to me.

elfosardo commented:
/approve just a nit, but I would probably change the title to "Allow disable disk cleaning" to avoid misunderstanding the scope of the change

elfosardo commented:
/test-integration

fmuyassarov commented:
/retitle Allow disable automated cleaning

fmuyassarov commented:
/test-integration 
Triggering CI even though it passed. I need to double check the functionality

fmuyassarov commented:
/test-integration

fmuyassarov commented:
/hold Until functional test are passing locally.

dtantsur commented on apis/metal3.io/v1alpha1/baremetalhost_types.go:
```diff
@@ -278,8 +278,25 @@ type BareMetalHostSpec struct {
+// Allowed automated cleaning modes
+const (
+	CleaningModeDisabled AutomatedCleaningMode = "Disabled"
+	CleaningModeFull     AutomatedCleaningMode = "Full"
```
Maybe not call it "Full" in case we ever support actually removing all data, not just the partitioning table? We in ironic call it "metadata cleaning".

fmuyassarov replied:
I can definitely rename it if desired. Any thoughts on this @dhellmann ?

dhellmann replied:
Future-proofing the name would be good.

If we're only cleaning metadata, but on all devices, should we reflect that? `MetadataAllDevices` seems a bit long but at least it's descriptive. We could then have `MetadataSelectedDevices` at some point and variations for `AllDataAllDevices` and `AllDataSelectedDevices` as well.

@dtantsur Is the switch to control whether we delete metadata or all data exposed in the Ironic API, or just a config setting?

Are we ever likely to support different modes per device?

fmuyassarov replied:
> Future-proofing the name would be good.
> 
> If we're only cleaning metadata, but on all devices, should we reflect that? `MetadataAllDevices` seems a bit long but at least it's descriptive. We could then have `MetadataSelectedDevices` at some point and variations for `AllDataAllDevices` and `AllDataSelectedDevices` as well.
> 
Maybe for now just `Metadata` ? Because, if we are going to extend it in the future, we will anyways have to touch the API. As such, maybe we could leave it to be done in future and for this proposal just stick to the `Metadata` ?

dhellmann replied:
OK, I can go along with `Metadata` applying to all devices, for now. And in the future we can have `MetadataSelectedDevices` or something like that.

fmuyassarov replied:
@dhellmann Updated the PR and the [proposal](https://github.com/metal3-io/metal3-docs/pull/166) as well.

dtantsur commented on pkg/provisioner/ironic/ironic.go:
```diff
@@ -1212,6 +1245,23 @@ func (p *ironicProvisioner) Provision(hostConf provisioner.HostConfigData) (resu
+		p.publisher("CleaningDisabled", "Automated cleaning disabled")
+	}
+
+	if ironicNode.AutomatedClean != nil {
```
It's `nil` by default. Not sure what you meant here, but it seems like it will never be updated?

fmuyassarov replied:
@dtantsur This was added here to avoid NPE. Removing `if ironicNode.AutomatedClean != nil` will cause runtime panic. (example [snippet](https://paste.ubuntu.com/p/MKXGcgGDZK/)). Maybe `ironicNode.AutomatedClean` in gophercloud should be defaulted to `True` as you mentioned during the review on the PR in gophercloud, or I need to assign `True` here before my logic comes into play.

fmuyassarov replied:
> It's `nil` by default. Not sure what you meant here, but it seems like it will never be updated?

It seems so.

fmuyassarov replied:
I did some code refactoring to verify if that's the issue, and indeed we are not able to update it just because ironicNode.AutomatedClean is `nil`

fmuyassarov replied:
As discussed on the [slack](https://kubernetes.slack.com/archives/CHD49TLE7/p1615195502083900), I changed the code a bit to enforce always updating the Ironic node.

dtantsur commented on pkg/provisioner/ironic/ironic.go:
```diff
@@ -1212,6 +1245,23 @@ func (p *ironicProvisioner) Provision(hostConf provisioner.HostConfigData) (resu
+	}
+
+	if ironicNode.AutomatedClean != nil {
+		if (p.host.Spec.AutomatedCleaningMode == metal3v1alpha1.CleaningModeDisabled && *ironicNode.AutomatedClean == true) || (p.host.Spec.AutomatedCleaningMode != metal3v1alpha1.CleaningModeDisabled && *ironicNode.AutomatedClean == false) {
```
Maybe move this check inside `updateAutomatedClean` to avoid duplication of logic?

dtantsur commented on pkg/provisioner/ironic/ironic.go:
```diff
@@ -1377,6 +1427,23 @@ func (p *ironicProvisioner) Deprovision(force bool) (result provisioner.Result,
+		p.publisher("CleaningDisabled", "Automated cleaning disabled")
+	}
+
+	if ironicNode.AutomatedClean != nil {
```
Same comments here

fmuyassarov commented:
/test-integration

fmuyassarov commented:
/hold cancel
@dtantsur @dhellmann @kashifest Please take look, it is ready from my side, I did a couple of test locally. The only open discussion that is unresolved is https://github.com/metal3-io/baremetal-operator/pull/784#discussion_r589422010, otherwise I think I addressed all the comments.

kashifest commented:
/lgtm

fmuyassarov commented:
/test-integration

dhellmann commented on pkg/provisioner/ironic/ironic.go:
```diff
@@ -354,6 +354,44 @@ func (p *ironicProvisioner) findExistingHost() (ironicNode *nodes.Node, err erro
+			"automated_clean", ironicNode.AutomatedClean,
+		)
+		p.publisher("CleaningDisabled", "Automated cleaning disabled")
+	} else if p.host.Spec.AutomatedCleaningMode != metal3v1alpha1.CleaningModeDisabled {
```
The condition here is redundant, isn't it?

Is there any reason not to always log what cleaning mode is being used? That would let us do something like the below and avoid the conditional entirely.

```go
p.log.Info("setting cleaning mode", 
    "ID", ironicNode.UUID, 
    "mode", p.host.Spec.AutomatedCleaningMode)
value = p.host.Spec.AutomatedCleaningMode != metal3v1alpha1.CleaningModeDisabled

updates = ...
```

fmuyassarov replied:
Yes it is! That's a good point, thanks.

dhellmann commented on pkg/provisioner/ironic/ironic.go:
```diff
@@ -1347,6 +1385,8 @@ func (p *ironicProvisioner) Provision(hostConf provisioner.HostConfigData) (resu
 		return transientError(provisioner.NeedsRegistration)
 	}
 
+	_ = p.updateAutomatedClean(ironicNode)
```
If the error was a 409, we need to pause before trying anything else because the host is busy. If it's not a 409, we probably don't want to ignore the error.

fmuyassarov replied:
Fixed it. Thanks.

dhellmann commented on pkg/provisioner/ironic/ironic.go:
```diff
@@ -1512,6 +1552,8 @@ func (p *ironicProvisioner) Deprovision(force bool) (result provisioner.Result,
 		return transientError(provisioner.NeedsRegistration)
 	}
 
+	_ = p.updateAutomatedClean(ironicNode)
```
We need to handle the error here, too.

fmuyassarov replied:
Yes, fixed it.

fmuyassarov commented on pkg/provisioner/ironic/provision_test.go:
```diff
@@ -66,7 +66,7 @@ func TestProvision(t *testing.T) {
 		{
 			name: "active state",
-			ironic: testserver.NewIronic(t).Ready().Node(nodes.Node{
+			ironic: testserver.NewIronic(t).WithDefaultResponses().Node(nodes.Node{
 				ProvisionState: string(nodes.Active),
 				UUID:           nodeUUID,
 			}),
```
@dhellmann Please take a look at this change. Because I'm not sure if this is okay.  I had to make change in order to pass unit tests. Once I added error handling as you [suggested](https://github.com/metal3-io/baremetal-operator/pull/784#discussion_r592578762) , unit tests started failing as below
```
$ make unit

   server.go:249: ironic: Cannot find any default response for [PATCH] /v1/nodes/33ce8659-7400-4c68-9535-d10766f07a58
    server.go:72: ironic: PATCH /v1/nodes/33ce8659-7400-4c68-9535-d10766f07a58 -> 
    --- FAIL: TestProvision/active_state (0.00s)
        server.go:17: inspector: new server created
        server.go:134: inspector: adding response for [GET] /v1
        server.go:134: inspector: adding response for [GET] /v1/introspection/33ce8659-7400-4c68-9535-d10766f07a58
        server.go:67: inspector: endpoint: http://127.0.0.1:41231/v1/
        provision_test.go:121: 
            	Error Trace:	provision_test.go:121
            	Error:      	Received unexpected error:
            	            	EOF
            	            	failed to update ironic node automated_clean
            	            	github.com/metal3-io/baremetal-operator/pkg/provisioner/ironic.(*ironicProvisioner).updateAutomatedClean
            	            		/home/fmuyassarov/go/src/github/baremetal-operator/pkg/provisioner/ironic/ironic.go:386
            	            	github.com/metal3-io/baremetal-operator/pkg/provisioner/ironic.(*ironicProvisioner).Provision
            	            		/home/fmuyassarov/go/src/github/baremetal-operator/pkg/provisioner/ironic/ironic.go:1386
            	            	github.com/metal3-io/baremetal-operator/pkg/provisioner/ironic.TestProvision.func1
            	            		/home/fmuyassarov/go/src/github/baremetal-operator/pkg/provisioner/ironic/provision_test.go:111
            	            	testing.tRunner
            	            		/usr/local/go/src/testing/testing.go:1194
            	            	runtime.goexit
            	            		/usr/local/go/src/runtime/asm_amd64.s:1371
            	            	could not update ironic node automated_clean field
            	            	github.com/metal3-io/baremetal-operator/pkg/provisioner/ironic.(*ironicProvisioner).Provision
            	            		/home/fmuyassarov/go/src/github/baremetal-operator/pkg/provisioner/ironic/ironic.go:1387
            	            	github.com/metal3-io/baremetal-operator/pkg/provisioner/ironic.TestProvision.func1
            	            		/home/fmuyassarov/go/src/github/baremetal-operator/pkg/provisioner/ironic/provision_test.go:111
            	            	testing.tRunner
            	            		/usr/local/go/src/testing/testing.go:1194
            	            	runtime.goexit
            	            		/usr/local/go/src/runtime/asm_amd64.s:1371
            	Test:       	TestProvision/active_state
```

```
$ make unit | grep PATCH
...
    server.go:72: ironic: PATCH /v1/nodes/33ce8659-7400-4c68-9535-d10766f07a58 -> {"automated_clean":null,"uuid":"33ce8659-7400-4c68-9535-d10766f07a58","name":"","power_state":"","target_power_state":"","provision_state":"","target_provision_state":"","maintenance":false,"maintenance_reason":"","fault":"","last_error":"","reservation":"","driver":"","driver_info":null,"driver_internal_info":null,"properties":null,"instance_info":null,"instance_uuid":"","chassis_uuid":"","extra":null,"console_enabled":false,"raid_config":null,"target_raid_config":null,"clean_step":null,"deploy_step":null,"resource_class":"","boot_interface":"","console_interface":"","deploy_interface":"","inspect_interface":"","management_interface":"","network_interface":"","power_interface":"","raid_interface":"","rescue_interface":"","storage_interface":"","traits":null,"vendor_interface":"","conductor_group":"","protected":false,"protected_reason":"","owner":""}
    server.go:249: ironic: Cannot find any default response for [PATCH] /v1/nodes/33ce8659-7400-4c68-9535-d10766f07a58
    server.go:72: ironic: PATCH /v1/nodes/33ce8659-7400-4c68-9535-d10766f07a58 -> 
    server.go:249: ironic: Cannot find any default response for [PATCH] /v1/nodes/33ce8659-7400-4c68-9535-d10766f07a58
    server.go:72: ironic: PATCH /v1/nodes/33ce8659-7400-4c68-9535-d10766f07a58 -> 
...
```

Replacing `testserver.NewIronic(t).Ready().Node` -> `testserver.NewIronic(t).WithDefaultResponses().Node` fixed the units tests. But I'm not 100% clear about those unit tests.

fmuyassarov replied:
I don't understand how my patch is breaking the existing unit tests.

dtantsur replied:
My guess: previously we would call PATCH only once for all operations. Now you've added another call, but there is no fixture for it.

hardys replied:
Do we actually have to do the PATCH in the Provision/Deprovision functions, or could we just add the update e.g to `getImageUpdateOptsForNode` ?

That would simplify the code changes, and reduce the test impact but perhaps I'm missing some corner cases which won't get handled in that case?

dtantsur replied:
When is it called? If it's only called before the deployment, it's too late to disable cleaning (it happens on the way from `manageable` to `available`).

fmuyassarov replied:
I'm not sure about `getImageUpdateOptsForNode`,  and the reason I used it in the Provision() & Deprovision() is because cleaning also can be enabled/disabled for provisioning & deprovisioning. I could not find any better place to call Ironic to update the field.

andfasano replied:
@fmuyassarov in case of problems the mock's output has been designed to help you arrange the required calls in the test setup: the most frequent reason it's a change in the code that stimulates one (or more) API calls.
The usage of `WithDefaultResponses` could be handy when, in the test setup, you're not interested in the "usual" API calls, and so its goal it's to add a valid (mostly generic) answer for such calls. On the other side, it could hide some important interactions that instead could be worth to be mocked. If you remove `WithDefaultResponses` and run the test again, the output should provide enough info to detect which call requires to be mocked (ie by looking at the messages in the format `... Cannot find any default response for ...`).

fmuyassarov replied:
Thank you @andfasano for the explanation! That's what I got now...

```
       server.go:72: ironic: GET /v1/nodes/33ce8659-7400-4c68-9535-d10766f07a58 -> {"automated_clean":null,"uuid":"33ce8659-7400-4c68-9535-d10766f07a58","name":"","power_state":"","target_power_state":"","provision_state":"active","target_provision_state":"","maintenance":false,"maintenance_reason":"","fault":"","last_error":"","reservation":"","driver":"","driver_info":null,"driver_internal_info":null,"properties":null,"instance_info":null,"instance_uuid":"","chassis_uuid":"","extra":null,"console_enabled":false,"raid_config":null,"target_raid_config":null,"clean_step":null,"deploy_step":null,"resource_class":"","boot_interface":"","console_interface":"","deploy_interface":"","inspect_interface":"","management_interface":"","network_interface":"","power_interface":"","raid_interface":"","rescue_interface":"","storage_interface":"","traits":null,"vendor_interface":"","conductor_group":"","protected":false,"protected_reason":"","owner":""}
    server.go:236: ironic: found default response for /v1/nodes/33ce8659-7400-4c68-9535-d10766f07a58: {200, {"automated_clean":null,"uuid":"{id}","name":"","power_state":"","target_power_state":"","provision_state":"","target_provision_state":"","maintenance":false,"maintenance_reason":"","fault":"","last_error":"","reservation":"","driver":"","driver_info":null,"driver_internal_info":null,"properties":null,"instance_info":null,"instance_uuid":"","chassis_uuid":"","extra":null,"console_enabled":false,"raid_config":null,"target_raid_config":null,"clean_step":null,"deploy_step":null,"resource_class":"","boot_interface":"","console_interface":"","deploy_interface":"","inspect_interface":"","management_interface":"","network_interface":"","power_interface":"","raid_interface":"","rescue_interface":"","storage_interface":"","traits":null,"vendor_interface":"","conductor_group":"","protected":false,"protected_reason":"","owner":""}}
    server.go:72: ironic: PATCH /v1/nodes/33ce8659-7400-4c68-9535-d10766f07a58 -> {"automated_clean":null,"uuid":"33ce8659-7400-4c68-9535-d10766f07a58","name":"","power_state":"","target_power_state":"","provision_state":"","target_provision_state":"","maintenance":false,"maintenance_reason":"","fault":"","last_error":"","reservation":"","driver":"","driver_info":null,"driver_internal_info":null,"properties":null,"instance_info":null,"instance_uuid":"","chassis_uuid":"","extra":null,"console_enabled":false,"raid_config":null,"target_raid_config":null,"clean_step":null,"deploy_step":null,"resource_class":"","boot_interface":"","console_interface":"","deploy_interface":"","inspect_interface":"","management_interface":"","network_interface":"","power_interface":"","raid_interface":"","rescue_interface":"","storage_interface":"","traits":null,"vendor_interface":"","conductor_group":"","protected":false,"protected_reason":"","owner":""}
    server.go:67: ironic: endpoint: http://127.0.0.1:39187/v1/
    server.go:72: ironic: GET /v1/nodes/33ce8659-7400-4c68-9535-d10766f07a58 -> {"automated_clean":null,"uuid":"33ce8659-7400-4c68-9535-d10766f07a58","name":"","power_state":"","target_power_state":"","provision_state":"cleaning","target_provision_state":"","maintenance":false,"maintenance_reason":"","fault":"","last_error":"","reservation":"","driver":"","driver_info":null,"driver_internal_info":null,"properties":null,"instance_info":null,"instance_uuid":"","chassis_uuid":"","extra":null,"console_enabled":false,"raid_config":null,"target_raid_config":null,"clean_step":null,"deploy_step":null,"resource_class":"","boot_interface":"","console_interface":"","deploy_interface":"","inspect_interface":"","management_interface":"","network_interface":"","power_interface":"","raid_interface":"","rescue_interface":"","storage_interface":"","traits":null,"vendor_interface":"","conductor_group":"","protected":false,"protected_reason":"","owner":""}
    server.go:249: ironic: Cannot find any default response for [PATCH] /v1/nodes/33ce8659-7400-4c68-9535-d10766f07a58
    server.go:72: ironic: PATCH /v1/nodes/33ce8659-7400-4c68-9535-d10766f07a58 -> 
    --- FAIL: TestProvision/other_state:_Cleaning (0.00s)
        server.go:17: inspector: new server created
        server.go:134: inspector: adding response for [GET] /v1
        server.go:134: inspector: adding response for [GET] /v1/introspection/33ce8659-7400-4c68-9535-d10766f07a58
        server.go:67: inspector: endpoint: http://127.0.0.1:37975/v1/
        provision_test.go:113: 
            	Error Trace:	provision_test.go:113
            	Error:      	Not equal: 
            	            	expected: true
            	            	actual  : false
            	Test:       	TestProvision/other_state:_Cleaning
        provision_test.go:114: 
            	Error Trace:	provision_test.go:114
            	Error:      	Not equal: 
            	            	expected: 10s
            	            	actual  : 0s
            	Test:       	TestProvision/other_state:_Cleaning
        provision_test.go:121: 
            	Error Trace:	provision_test.go:121
            	Error:      	Received unexpected error:
            	            	EOF
            	            	failed to update ironic node automated_clean
            	            	github.com/metal3-io/baremetal-operator/pkg/provisioner/ironic.(*ironicProvisioner).updateAutomatedClean
            	            		/home/fmuyassarov/go/src/github/baremetal-operator/pkg/provisioner/ironic/ironic.go:386
            	            	github.com/metal3-io/baremetal-operator/pkg/provisioner/ironic.(*ironicProvisioner).Provision
            	            		/home/fmuyassarov/go/src/github/baremetal-operator/pkg/provisioner/ironic/ironic.go:1386
            	            	github.com/metal3-io/baremetal-operator/pkg/provisioner/ironic.TestProvision.func1
            	            		/home/fmuyassarov/go/src/github/baremetal-operator/pkg/provisioner/ironic/provision_test.go:111
            	            	testing.tRunner
            	            		/usr/local/go/src/testing/testing.go:1194
            	            	runtime.goexit
            	            		/usr/local/go/src/runtime/asm_amd64.s:1371
            	            	could not update ironic node automated_clean field
            	            	github.com/metal3-io/baremetal-operator/pkg/provisioner/ironic.(*ironicProvisioner).Provision
            	            		/home/fmuyassarov/go/src/github/baremetal-operator/pkg/provisioner/ironic/ironic.go:1387
            	            	github.com/metal3-io/baremetal-operator/pkg/provisioner/ironic.TestProvision.func1
            	            		/home/fmuyassarov/go/src/github/baremetal-operator/pkg/provisioner/ironic/provision_test.go:111
            	            	testing.tRunner
            	            		/usr/local/go/src/testing/testing.go:1194
            	            	runtime.goexit
            	            		/usr/local/go/src/runtime/asm_amd64.s:1371
            	Test:       	TestProvision/other_state:_Cleaning
```

andfasano replied:
To fix it, as also @dtantsur suggested, you need to instruct the Ironic mock to give a valid answer for a node PATCH using `NodeUpdate`, ie:

```
              {
			name: "active state",
			ironic: testserver.NewIronic(t).Ready().Node(nodes.Node{
				ProvisionState: string(nodes.Active),
				UUID:           nodeUUID,
			}).NodeUpdate(nodes.Node{
				UUID: nodeUUID,
			}),
			expectedRequestAfter: 0,
			expectedDirty:        false,
		},
```

fmuyassarov replied:
Great, that indeed fixed my issue! Thank you @andfasano ! Please take a look.

dtantsur commented on apis/metal3.io/v1alpha1/baremetalhost_types.go:
```diff
@@ -344,8 +344,25 @@ type BareMetalHostSpec struct {
+// Allowed automated cleaning modes
+const (
+	CleaningModeDisabled AutomatedCleaningMode = "Disabled"
+	CleaningModeMetadata AutomatedCleaningMode = "Metadata"
```
Thanks! There is no API knob to tune which cleaning runs, but there may be in the future.

dtantsur commented on pkg/provisioner/ironic/ironic.go:
```diff
@@ -1347,6 +1383,10 @@ func (p *ironicProvisioner) Provision(hostConf provisioner.HostConfigData) (resu
 	}
 
+	if err = p.updateAutomatedClean(ironicNode); err != nil {
+		return transientError(errors.Wrap(err, "could not update ironic node automated_clean field"))
```
nit: I'd prefer this to be phrased in terms of metal3 entities, not ironic fields

fmuyassarov replied:
Sure, will update it.

fmuyassarov commented:
Thank you all for your reviews. I think I have addressed all the comments I had so far. Please take a look.

andfasano commented on pkg/provisioner/ironic/ironic.go:
```diff
@@ -354,6 +354,42 @@ func (p *ironicProvisioner) findExistingHost() (ironicNode *nodes.Node, err erro
+		},
+	}
+
+	_, err = nodes.Update(p.client, ironicNode.UUID, updates).Extract()
```
Wouldn't be better to request an update only _if_ required? Otherwise the call will be raised everytime (and the current changes on the unit tests are showing that)

dtantsur replied:
Knowing whether it's required can be tricky. The default value is `nil`, which means "use the default from ironic.conf". So you'll have to assume what's in ironic.conf.

fmuyassarov replied:
> Wouldn't be better to request an update only _if_ required? Otherwise the call will be raised everytime (and the current changes on the unit tests are showing that)

It would be, but `ironicNode.AutomatedClean` has nil by default as @dtantsur pointed out. As such, for example adding this:
```
if p.host.Spec.AutomatedCleaningMode == metal3v1alpha1.CleaningModeDisabled && *ironicNode.AutomatedClean != false || p.host.Spec.AutomatedCleaningMode == metal3v1alpha1.CleaningModeMetadata && *ironicNode.AutomatedClean == false {
    // do the update here
}
```
will cause runtime panic issue as we are using `*ironicNode.AutomatedClean`.
```
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x1484c44]
```
 We had the [slack chat](https://kubernetes.slack.com/archives/CHD49TLE7/p1615195502083900) about this with @dtantsur, and thought the way forward would be to always update it.

andfasano replied:
Of course dereferencing a pointer without a nil check it's not feasible. And I don't think it's a blocker, considering that the call it's performed just in the provisioning/deprovisioning, but it would be nice to have a way to use that flag.

>The default value is nil, which means "use the default from ironic.conf". So you'll have to assume what's in ironic.conf.
Any way to retrieve the global conf in that case?

dtantsur replied:
There is no way to retrieve the global conf. What we could do is to avoid updates after the flag has been updated once (and its value is not `nil`).

fmuyassarov commented:
@andfasano @dhellmann @dtantsur @hardys 
I think all the comments have been addressed by now. Also the [update](https://github.com/metal3-io/metal3-docs/pull/166) in the original proposal got merged. Can we merge this one too if looks okay? Because I will still need to update it in [CAPM3 PR ](https://github.com/metal3-io/cluster-api-provider-metal3/pull/156)after this is merged.

hardys commented:
> @andfasano @dhellmann @dtantsur @hardys
> I think all the comments have been addressed by now. Also the [update](https://github.com/metal3-io/metal3-docs/pull/166) in the original proposal got merged. Can we merge this one too if looks okay? Because I will still need to update it in [CAPM3 PR ](https://github.com/metal3-io/cluster-api-provider-metal3/pull/156)after this is merged.

This version lgtm - I'll trigger the integration tests and do some local testing of the Disabled case before approving (and hopefully we can get some other reviews in the meantime).

hardys commented:
/test-integration

zaneb commented on pkg/provisioner/ironic/ironic.go:
```diff
@@ -1347,6 +1383,10 @@ func (p *ironicProvisioner) Provision(hostConf provisioner.HostConfigData) (resu
 		return transientError(provisioner.NeedsRegistration)
 	}
 
+	if err = p.updateAutomatedClean(ironicNode); err != nil {
```
It seems like a mistake to call this in Provision() (or even Deprovision()), since there are numerous opportunities for the node to get cleaned at other times - notably right after registration, but also at any other time when the ironic database is dropped and we don't call Adopt().
I think we should be setting the cleaning mode in ValidateManagementAccess(), which gets called on every reconcile. That way the node is always in a state with the right setting. Perhaps we could consolidate this along with the 3 calls we are already making to nodes.Update() in that method into a single one.

fmuyassarov replied:
I remember a similar [comment](https://github.com/metal3-io/metal3-docs/pull/151#discussion_r527709842) from @dhellmann on the proposal that we could to do it in `ValidateManagementAccess()`, but I did it within `Provision()` and  `Deprovision()` because I thought these are the only methods from after which Ironic node gets cleaned. From Ironic [state machine](https://docs.openstack.org/ironic/pike/_images/states.svg). I see that Ironic cleaning state will come to cleaning state either after managable (BMH Ready state) or active -> deleting (BMH deprovisioning state).
@zaneb Would you mind to share an example use case where Ironic node gets cleaned without us calling `Provision()` or  `Deprovision()` ?

zaneb replied:
You're correct, I misremembered the change in 18465130c8f42bda4aa8191b5093656dcc029d90.
I do plan to change the state machine so that the Preparing state will move the state from Manageable->Available (and thus trigger automatic cleaning) as [proposed on the mailing list](https://groups.google.com/g/metal3-dev/c/N-EpqQn94bw/m/nRkM9bmuAAAJ), but currently it does not do so.

hardys replied:
I made a similar [comment](https://github.com/metal3-io/baremetal-operator/pull/784#discussion_r594475109) but the current approach works, so perhaps we can address the consolidation of Updates in a follow-up PR?

zaneb commented on apis/metal3.io/v1alpha1/baremetalhost_types.go:
```diff
@@ -344,8 +344,25 @@ type BareMetalHostSpec struct {
+
+// Allowed automated cleaning modes
+const (
+	CleaningModeDisabled AutomatedCleaningMode = "Disabled"
```
Other constants in the API are lower-case, should we do that here too?

fmuyassarov replied:
Yes, and this means allowed values will also need to be changed in the baremetalhost types. But your point makes sense. Will update it.

zaneb commented on docs/api.md:
```diff
@@ -172,6 +172,12 @@ The sub-fields are
+
+An interface to enable/disable automated cleaning during provisioning
+and deprovisioning. When set to `Disabled`, automated cleaning will be
+skipped, where `Metadata`(default value) enables it.
```
How do we expect 'Metadata' to differ from other possible cleaning modes?

fmuyassarov replied:
This was decided to be called Metadata to reflect Ironic Metadata cleaning after the discussion [here](https://github.com/metal3-io/baremetal-operator/pull/784#discussion_r589422010). So, in the future there might be on option to select only certain devices to be cleaned.

zaneb commented on pkg/provisioner/ironic/ironic.go:
```diff
@@ -354,6 +354,42 @@ func (p *ironicProvisioner) findExistingHost() (ironicNode *nodes.Node, err erro
 
+func (p *ironicProvisioner) updateAutomatedClean(ironicNode *nodes.Node) (err error) {
+	var updates nodes.UpdateOpts
+	var value bool
```
nit: these declarations aren't needed, you can just use `:=`

fmuyassarov commented:
/test-integration

fmuyassarov commented:
/test govet

fmuyassarov commented:
/test generate

fmuyassarov commented:
@hardys previously integration CI passed, but I had some comments to be fixed from @zaneb, and now it is ready again. I can trigger it now if we are good to go with the current state. Please let me know If I can help in testing the feature.

hardys commented:
/test-integration

hardys commented:
This lgtm - I tested it locally and it works as expected - I see @zaneb made a comment similar to our previous discussion about when to handle the `nodes.Update` - we probably can improve/consolidate that but perhaps that can be done as a followup or as part of the future cleaning rework mentioned by @zaneb ?

/lgtm

fmuyassarov commented:
> This lgtm - I tested it locally and it works as expected - I see @zaneb made a comment similar to our previous discussion about when to handle the `nodes.Update` - we probably can improve/consolidate that but perhaps that can be done as a followup or as part of the future cleaning rework mentioned by @zaneb ?
> 
> /lgtm

Thanks @hardys ! If possible, it would be nice if we do it in a follow up PR, and I can sign up to help on that.

andfasano commented:
/lgtm

fmuyassarov commented:
@dhellmann would you mind to take a look once more please?

zaneb commented on pkg/provisioner/ironic/ironic.go:
```diff
@@ -354,6 +354,39 @@ func (p *ironicProvisioner) findExistingHost() (ironicNode *nodes.Node, err erro
+	case nil:
+	case gophercloud.ErrDefault409:
+		p.log.Info("could not update ironic node automated_clean, busy")
+		_, err = retryAfterDelay(provisionRequeueDelay)
```
This line just returns nil and does not retry. This is wrong and will cause the calling code to believe that cleaning has been disabled when it hasn't.

fmuyassarov replied:
@zaneb you are right. Do you think I should not retry at all, and replace L380 with for example something like
`return errors.Wrap(err, "could not update ironic node automated_clean, busy")` or shall I instead add `result provisioner.Result` as second return value for `updateAutomatedClean()` and do retry? Maybe the latter option makes more sense, because I can see that we are retrying in most of the places in ironic.go.

zaneb replied:
Returning an `error` will result in an immediate* requeue plus an ugly error in the logs. Generally speaking, responding to ironic getting hammered by hammering it harder is not ideal, which is why we use retryAfterDelay everywhere else we get a 409.

\* subject to throttling in the controller

fmuyassarov replied:
Fixed accordingly, thank you!

zaneb commented on pkg/provisioner/ironic/ironic.go:
```diff
@@ -354,6 +354,39 @@ func (p *ironicProvisioner) findExistingHost() (ironicNode *nodes.Node, err erro
+		_, err = retryAfterDelay(provisionRequeueDelay)
+		return
+	default:
+		_, err = transientError(errors.Wrap(err, "failed to update ironic node automated_clean"))
```
Using the transientError call here is pointless in a function that only returns an `error`.

fmuyassarov replied:
True, will fix it once we decide on the questions above.

zaneb reviewed (commented):
OK, if someone is going to follow up with moving it to the registration then I'm OK with doing it later.

That said, there's a bug in the error handling here and it might be quicker to move the code than to fix it :shrug:

dtantsur commented:
/lgtm cancel

@fmuyassarov could you update per Zane's comments?

dtantsur commented:
/approve

Approving since the direction and the overall code is correct, just some details need polishing.

flaper87 commented on pkg/provisioner/ironic/ironic.go:
```diff
@@ -354,6 +354,39 @@ func (p *ironicProvisioner) findExistingHost() (ironicNode *nodes.Node, err erro
+		"ID", ironicNode.UUID,
+		"mode", p.host.Spec.AutomatedCleaningMode)
+
+	p.publisher("CleaningMode", string(p.host.Spec.AutomatedCleaningMode))
```
Not sure if there's a problem in my env but this resulted being quite verbose for me:

![DeepinScreenshot_select-area_20210324082825](https://user-images.githubusercontent.com/13816/112278594-eae97800-8c7a-11eb-959b-0381c8ae709e.png)

fmuyassarov replied:
Seems that's due the publisher is being called on all the states if Ironic node. 
@hardys do you think it is okay that I add the publisher (for deprovisioning) when Ironic node is either in [`Active or DeployFail`](https://github.com/metal3-io/baremetal-operator/blob/5ebf0ab6e367a854c4177edc6612381b1d65dc01/pkg/provisioner/ironic/ironic.go#L1600)  or maybe when it is [`deleting`](https://github.com/metal3-io/baremetal-operator/blob/5ebf0ab6e367a854c4177edc6612381b1d65dc01/pkg/provisioner/ironic/ironic.go#L1584) ?

And during provisioning, I was thinking maybe to add publisher when Ironic node is in [Active](https://github.com/metal3-io/baremetal-operator/blob/5ebf0ab6e367a854c4177edc6612381b1d65dc01/pkg/provisioner/ironic/ironic.go#L1477)?

hardys replied:
@fmuyassarov do we need an event for this at all?

I'm thinking we don't output events for many other types of configuration, so we could just move this update to `ValidateManagementAccess()` as @zaneb mentioned, and apend it to the [other updates](https://github.com/metal3-io/baremetal-operator/blob/master/pkg/provisioner/ironic/ironic.go#L472..L479)?  Maybe log something but drop the event output?

fmuyassarov replied:
I added events here as per https://github.com/metal3-io/baremetal-operator/pull/784#discussion_r585064586. I can replace events with simple logging if that's okay.

fmuyassarov replied:
I have removed events, and instead kept only logging that was on top of that. I hope that's ok ?
```
	p.log.Info("setting automated cleaning mode to",
		"ID", ironicNode.UUID,
		"mode", p.host.Spec.AutomatedCleaningMode)
```

fmuyassarov replied:
@hardys  Isn't `ValidateManagementAccess` is used only when we are registering the host for the first time? If yes, that will work when someone enables/disables cleaning before provisioning. But what if we want to enable/disable cleaning before deprovisioning?

fmuyassarov replied:
@zaneb any thoughts on above question?

zaneb replied:
`ValidateManagementAccess()` is called in ~every state since #716.

hardys replied:
> I added events here as per [#784 (comment)](https://github.com/metal3-io/baremetal-operator/pull/784#discussion_r585064586). I can replace events with simple logging if that's okay.

Logging wfm - my previous comment was just to say that I did generate an event in the inspection-disabled case - but if we generate an event (or log output) we want to do that once, not repeatedly

fmuyassarov commented:
/test-integration

hardys commented on pkg/provisioner/ironic/ironic.go:
```diff
@@ -546,6 +546,36 @@ func (p *ironicProvisioner) ValidateManagementAccess(credentialsChanged, force b
+	if (p.host.Spec.AutomatedCleaningMode == "disabled" && *ironicNode.AutomatedClean != false) || (p.host.Spec.AutomatedCleaningMode == "enabled" && *ironicNode.AutomatedClean == false) {
+		value := p.host.Spec.AutomatedCleaningMode != metal3v1alpha1.CleaningModeDisabled
+
+		updates := nodes.UpdateOpts{
```
Does this need to be an additional update?

I was thinking this could be called immediately after `getImageUpdateOptsForNode` (e.g a new `getCleaningModeOpts ` function perhaps?) and appended to the `updates`?

@zaneb am I right in saying we should hit that code in both the Provisioning and Deprovisioning state, since imageData will be set via either the Spec or Status?

zaneb replied:
I [agree](https://github.com/metal3-io/baremetal-operator/pull/784#discussion_r597888586), we should stop copying and pasting the code to Update() and all of its error handling, which we already call in 3 places (now 4) when 1 would suffice, and just add one more thing to the list.

Note, however, that the `getImageUpdateOptsForNode()` is only called if the node doesn't exist. Which is a bug, because if updating it fails then it will never get called again. You are correct that the image data is available in both the Provisioning and Deprovisioning states.

We could refactor and fix the bug in a separate PR if that helps.

fmuyassarov replied:
Do we need a new func for this? or can we just add it within  `getImageUpdateOptsForNode()?`

zaneb replied:
I would just do the Update() once after the if/else, and have the stuff that is currently calling Update() just add stuff to the updates list.
We should definitely not call Update() within `getImageUpdateOptsForNode()` with the same list of updates that we're about to return, only for Update() to be called with it again.

fmuyassarov replied:
@zaneb can you please take a look if the current state is what you meant?

zaneb replied:
I wouldn't put it inside `getImageUpdateOptsForNode()` because:

1. of the bug mentioned above, which means it is not called most of the time when we want it to be called (you'd have to delete the ironic DB to get the cleaning mode to update)
2. even when the bug is fixed, it'll only be called when there is an image specified, but that's not a condition we want to impose on whether the cleaning mode gets updated

hardys commented on pkg/provisioner/ironic/ironic.go:
```diff
@@ -546,6 +546,36 @@ func (p *ironicProvisioner) ValidateManagementAccess(credentialsChanged, force b
 
+	p.log.Info("setting automated cleaning mode to",
+		"ID", ironicNode.UUID,
+		"mode", p.host.Spec.AutomatedCleaningMode)
```
We should only log this when actually making a change

zaneb commented on pkg/provisioner/ironic/ironic.go:
```diff
@@ -391,7 +391,7 @@ func (p *ironicProvisioner) ValidateManagementAccess(credentialsChanged, force b
 	driverInfo["deploy_ramdisk"] = deployRamdiskURL
 
-	result, err = operationComplete()
+	result, _ = operationComplete()
```
This change doesn't do anything and no good can come of it.

fmuyassarov commented:
@zaneb @hardys do you think it is better now?

hardys commented on pkg/provisioner/ironic/ironic.go:
```diff
@@ -508,61 +500,61 @@ func (p *ironicProvisioner) ValidateManagementAccess(credentialsChanged, force b
+			Op:    nodes.ReplaceOp,
+			Path:  "/automated_clean",
+			Value: value,
+		},
```
This UpdateOperation is still applied unconditionally, should we be checking `ironicNode.AutomatedClean` and only setting it while needed?

zaneb replied:
Yeah, you also need to check for a non-nil pointer before dereferencing it.

fmuyassarov replied:
> Yeah, you also need to check for a non-nil pointer before dereferencing it.

We will never be able to do the update if we check that, because `automated_clean` = `nil` by default in gophercloud.

zaneb replied:
If it's nil you always update:

```
if ironicNode.AutomatedClean == nil || (p.host.Spec.AutomatedCleaningMode == "disabled" && *ironicNode.AutomatedClean != false) || (p.host.Spec.AutomatedCleaningMode == "enabled" && *ironicNode.AutomatedClean == false) {
```

fmuyassarov replied:
🤦‍♂️, thank you!

fmuyassarov commented on pkg/provisioner/ironic/ironic.go:
```diff
@@ -508,61 +500,61 @@ func (p *ironicProvisioner) ValidateManagementAccess(credentialsChanged, force b
+			Value: value,
+		},
+	)
+	ironicNode, err = nodes.Update(p.client, ironicNode.UUID, updates).Extract()
```
I think yes, with the current state.

hardys replied:
Ok if we think that can happen we probably need to check len(updates) and only do the nodes.Update if there are updates pending?

hardys replied:
nvm, I see the latest version does that :)

hardys commented:
This version lgtm other than resolving the failing tests, thanks @fmuyassarov!

fmuyassarov commented:
/test-integration

fmuyassarov commented:
Thanks everyone for reviews. I think all the comments now addressed!.

zaneb commented:
/lgtm
